activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb

```diff
@@ -91,6 +96,16 @@ def sql_for_insert(sql, pk, binds, returning) # :nodoc:
         end
         private :sql_for_insert
 
+        def sql_for_update(sql, binds, returning_columns) # :nodoc:
+          return super unless returning_columns&.any?
+
+          returning_columns_statement = returning_columns.map { |c| quote_column_name(c) }.join(", ")
+          sql = "#{sql} RETURNING #{returning_columns_statement}"
```

Comment:
Instead of concat sqls, can we just build the right AST in the right place? We probably need to build the AST before passing to `exec_update` in the `update` method.

Reply:
Sorry I'm not entirely clear on the ask here. Is there something different here than `sql_for_insert` which uses the same code pattern?

I'm planning to move this method from being postgres specific and instead put it under the abstract class.

Reply:
My two cents is any further changes around this would be better handled as a separate follow-up PR? A refactor really ought to also include changes to `sql_for_insert`. However, including changes to that method here would be too broad and make this PR unfocused.